domain/network/state/space.go

```diff
@@ -0,0 +1,292 @@
+func (st *State) GetSpaceByName(
+	ctx context.Context,
+	name string,
+) (*network.SpaceInfo, error) {
+	db, err := st.DB()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	// Append the space.name condition to the query.
+	q := retrieveSpacesStmt + " WHERE space.name = $M.name;"
+
+	s, err := sqlair.Prepare(q, Space{}, sqlair.M{})
+	if err != nil {
+		return nil, errors.Annotatef(err, "preparing %q", q)
+	}
+
+	var rows Spaces
+	if err := db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		return errors.Trace(tx.Query(ctx, s, sqlair.M{"name": name}).GetAll(&rows))
+	}); err != nil {
+		return nil, errors.Annotate(err, "querying spaces by name")
+	}
+
+	if len(rows) == 0 {
+		return nil, errors.NotFoundf("space name %q", name)
+	}
+
+	return &rows.ToSpaceInfos()[0], nil
+}
+
+// GetAllSpaces returns all spaces for the model.
+func (st *State) GetAllSpaces(
+	ctx context.Context,
+) (network.SpaceInfos, error) {
+
+	db, err := st.DB()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	s, err := sqlair.Prepare(retrieveSpacesStmt, Space{})
+	if err != nil {
+		return nil, errors.Annotatef(err, "preparing %q", retrieveSpacesStmt)
+	}
+
+	var rows Spaces
+	if err := db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		return errors.Trace(tx.Query(ctx, s).GetAll(&rows))
+	}); err != nil {
+		return nil, errors.Annotate(err, "querying all spaces")
+	}
+
+	return rows.ToSpaceInfos(), nil
+}
+
+// UpdateSpace updates the space identified by the passed uuid.
+func (st *State) UpdateSpace(
+	ctx context.Context,
+	uuid string,
+	name string,
+) error {
+	db, err := st.DB()
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	q := `
+UPDATE space
+SET    name = ?
+WHERE  uuid = ?;`
+	return db.StdTxn(ctx, func(ctx context.Context, tx *sql.Tx) error {
+		if _, err := tx.ExecContext(ctx, q, name, uuid); err != nil {
+			return errors.Trace(err)
+		}
+		return nil
+	})
+}
+
+// DeleteSpace deletes the space identified by the passed uuid.
+func (st *State) DeleteSpace(
+	ctx context.Context,
+	uuid string,
+) error {
+	db, err := st.DB()
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	deleteSpaceStmt := "DELETE FROM space WHERE uuid = ?;"
+	deleteProviderSpaceStmt := "DELETE FROM provider_space WHERE space_uuid = ?;"
+	updateSubnetSpaceUUIDStmt := "UPDATE subnet SET space_uuid = ? WHERE space_uuid = ?;"
+
+	return db.StdTxn(ctx, func(ctx context.Context, tx *sql.Tx) error {
+		delProviderSpaceResult, err := tx.ExecContext(ctx, deleteProviderSpaceStmt, uuid)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		delProviderSpaceAffected, err := delProviderSpaceResult.RowsAffected()
+		if err != nil {
+			return errors.Trace(err)
+		}
+		if delProviderSpaceAffected != 1 {
+			return fmt.Errorf("provider space id for space %s not found", uuid)
+		}
+
+		if _, err := tx.ExecContext(ctx, updateSubnetSpaceUUIDStmt, network.AlphaSpaceId, uuid); err != nil {
+			return errors.Trace(err)
+		}
```

Comment:
Check the result info.

Reply:
here we don't need to, we must be able to delete a space that's not linked to any subnet